### PR TITLE
class instance dispatch using cascading and caching

### PR DESCRIPTION
**Pull request by yohto**
### What is the necessity of this update? What is updated?
Cascading check instance of class needs if-then-else if structure then it needs O( n ) time.
This PR method uses cascading check first, and then, caches the class and dispatched task.

### How did you do the test? (Not required for updating documents)
Create test for cascading dispatch.

### CODE REVIEW

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/json/JacksonArrayFormatter.java [hunk 1]
```diff
@@ -25,29 +25,40 @@
...
+    sw.setDefault((array, childObj) -> array.add(ObjectToJsonNode.get(childObj)));
+    sw.set(List.class, (array, childObj) ->
+        array.add(JacksonContainerToJsonObject.getFromList((List<Object>)childObj)));
+    sw.set(Map.class, (array, childObj) ->
+        array.add(JacksonContainerToJsonObject.getFromMap((Map<Object,Object>)childObj)));
+    dispatcher = sw.create();
+  }
 
   @Override
   public JsonNode write( final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/json/JacksonArrayFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/json/JacksonObjectFormatter.java [hunk 2]
```diff
@@ -25,34 +25,44 @@
...
+        objectNode.put(key, ObjectToJsonNode.get(childObj)));
+    sw.set(List.class, (objectNode, key, childObj) ->
+        objectNode.put(key, JacksonContainerToJsonObject.getFromList((List<Object>)childObj)));
+    sw.set(Map.class, (objectNode, key, childObj) ->
+        objectNode.put(key, JacksonContainerToJsonObject.getFromMap((Map<Object,Object>)childObj)));
+    dispatcher = sw.create();
+  }
 
   @Override
   public JsonNode write( final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/json/JacksonObjectFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextBooleanFormatter.java [hunk 3]
```diff
@@ -21,25 +21,35 @@
...
+        ((Boolean)obj).toString().getBytes("UTF-8")));
+    sw.set(String.class, (buffer, obj) -> buffer.append(
+        Boolean.valueOf("true".equals((String)obj)).toString().getBytes("UTF-8")));
+    sw.set(PrimitiveObject.class, (buffer, obj) -> buffer.append(
+        Boolean.valueOf(((PrimitiveObject)obj).getBoolean()).toString().getBytes("UTF-8")));
+    dispatcher = sw.create();
+  }
 
   @Override
   public void write( final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextBooleanFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextBytesFormatter.java [hunk 4]
```diff
@@ -21,19 +21,34 @@
...
+    sw.set(byte[].class, (buffer, obj) -> buffer.append((byte[])obj));
+    sw.set(String.class, (buffer, obj) -> buffer.append(((String)obj).getBytes("UTF-8")));
+    sw.set(PrimitiveObject.class, (buffer, obj) ->
+        buffer.append(((PrimitiveObject)obj).getBytes()));
+    dispatcher = sw.create();
+  }
+
 
   @Override
-  public void write(final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextBytesFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextDoubleFormatter.java [hunk 5]
```diff
@@ -21,34 +21,56 @@
...
+    dispatcher = sw.create();
+  }
 
-  private byte[] convert( final double target ) throws IOException {
+  private static byte[] convert(final double target) throws IOException {
     return Double.valueOf( target ).toString().getBytes("UTF-8");
   }
 
   @Override
   public void write( final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextDoubleFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextFloatFormatter.java [hunk 6]
```diff
@@ -21,34 +21,56 @@
...
+    dispatcher = sw.create();
+  }
 
-  private byte[] convert( final float target ) throws IOException {
+  private static byte[] convert( final float target ) throws IOException {
     return Float.valueOf( target ).toString().getBytes("UTF-8");
   }
 
   @Override
-  public void write(final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextFloatFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextFormatterFactory.java [hunk 7]
```diff
@@ -31,42 +31,43 @@
...
+    sw.set(ShortField.class,   schema -> new TextShortFormatter());
+    sw.set(StringField.class,  schema -> new TextStringFormatter());
+    sw.set(NullField.class,    schema -> new TextNullFormatter());
+    dispatcher = sw.create();
+  }
 
   /**
    * Create an ITextFormatter from a schema.
    */
-  public static ITextFormatter get( final IField schema ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextFormatterFactory.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextIntegerFormatter.java [hunk 8]
```diff
@@ -21,34 +21,56 @@
...
+    dispatcher = sw.create();
+  }
 
-  private byte[] convert( final int target ) throws IOException {
+  private static byte[] convert( final int target ) throws IOException {
     return Integer.valueOf( target ).toString().getBytes("UTF-8");
   }
 
   @Override
   public void write(final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextIntegerFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextLongFormatter.java [hunk 9]
```diff
@@ -21,34 +21,56 @@
...
+    dispatcher = sw.create();
+  }
 
-  private byte[] convert( final long target ) throws IOException {
+  private static byte[] convert( final long target ) throws IOException {
     return Long.valueOf( target ).toString().getBytes("UTF-8");
   }
 
   @Override
-  public void write( final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextLongFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextShortFormatter.java [hunk 10]
```diff
@@ -21,34 +21,57 @@
...
+  }
+
 
-  private byte[] convert( final short target ) throws IOException {
+  private static byte[] convert( final short target ) throws IOException {
     return Short.valueOf( target ).toString().getBytes("UTF-8");
   }
 
   @Override
-  public void write(final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextShortFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextStringFormatter.java [hunk 11]
```diff
@@ -21,19 +21,33 @@
...
+    ObjectDispatchByClass<DispatchedFunc> sw = new ObjectDispatchByClass<>();
+    sw.set(byte[].class, (buffer, obj) -> buffer.append((byte[])obj));
+    sw.set(String.class, (buffer, obj) -> buffer.append(((String)obj).getBytes("UTF-8")));
+    sw.set(PrimitiveObject.class, (buffer, obj) ->
+        buffer.append(((PrimitiveObject)obj).getBytes()));
+    dispatcher = sw.create();
+  }
 
   @Override
-  public void write( final ByteArrayData buffer , final Object obj ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/formatter/text/TextStringFormatter.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/objects/JavaObjectToPrimitiveObject.java [hunk 12]
```diff
@@ -18,36 +18,38 @@
...
+    sw.set(String.class,  obj -> new StringObj((String)obj));
+    dispatcher = sw.create();
+  }
 
   private JavaObjectToPrimitiveObject() {}
 
   /**
    * Convert Java objects to PrimitiveObject.
    */
-  public static PrimitiveObject get( final Object obj ) {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/objects/JavaObjectToPrimitiveObject.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/objects/JsonNodeToPrimitiveObject.java [hunk 13]
```diff
@@ -31,39 +31,41 @@
...
-    } else if ( jsonNode instanceof POJONode ) {
-      return new BytesObj( ( (POJONode)jsonNode ).binaryValue() );
-    } else if ( jsonNode instanceof NullNode ) {
-      return NullObj.getInstance();
-    } else if ( jsonNode instanceof MissingNode ) {
-      return NullObj.getInstance();
-    } else {
-      return new StringObj( jsonNode.toString() );
-    }
+  public static PrimitiveObject get(final JsonNode jsonNode) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/objects/JsonNodeToPrimitiveObject.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/objects/ObjectToJsonNode.java [hunk 14]
```diff
@@ -27,37 +27,39 @@
...
-    } else if ( obj instanceof Double ) {
-      return new DoubleNode( (Double)obj );
-    } else if ( obj instanceof byte[] ) {
-      return new BinaryNode( (byte[])obj );
-    } else if ( obj == null ) {
-      return NullNode.getInstance();
-    } else {
-      return new TextNode( obj.toString() );
-    }
+  public static JsonNode get(final Object obj) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/objects/ObjectToJsonNode.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/parser/java/JavaParserFactory.java [hunk 15]
```diff
@@ -19,24 +19,32 @@
...
+    sw.set(Map.class,  obj -> new JavaMapParser((Map)obj));
+    dispatcher = sw.create();
+  }
 
   private JavaParserFactory() {}
 
   /**
    * Create IParser from Java objects.
    */
-  public static IParser get( final Object obj ) {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/parser/java/JavaParserFactory.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/message/parser/json/JsonNodeToParser.java [hunk 16]
```diff
@@ -23,32 +23,46 @@
...
+      .setDefault(false)
+      .set(ObjectNode.class, true)
+      .set(ArrayNode.class, true)
+      .create();
+  }
 
   /**
    * Convert JsonNode to IParser.
    */
-  public static IParser get( final JsonNode jsonNode ) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/message/parser/json/JsonNodeToParser.java]: Create test cases of all types.

File: src/main/java/jp/co/yahoo/yosegi/spread/column/ColumnTypeFactory.java [hunk 17]
```diff
@@ -53,57 +87,15 @@ private ColumnTypeFactory() {}
...
-        return ColumnType.EMPTY_SPREAD;
-      }
-      if ( parser.isArray() ) {
-        return ColumnType.ARRAY;
-      }
-      return ColumnType.SPREAD;
-    }
-
-    if ( obj == null ) {
+  public static ColumnType get(final Object obj) throws IOException {
```
Comment (koijima) [inline src/main/java/jp/co/yahoo/yosegi/spread/column/ColumnTypeFactory.java]: Create test cases of all types.

**Comment by yohto** [comment]
Thank you for your review. I restrict this PR first. I will PR again after adding the more test cases later.

**Pull request closed without merging**